### PR TITLE
More TR natives

### PR DESCRIPTION
**Pull request by SlidyBat**
Adds some more TR natives that expose more properties of the trace results. Here's a short summary of the new natives:
```
TR_GetStartPosition - Self explanatory, added this in just for completion since GetEndPosition exists.
TR_GetFractionLeftSolid - Similar to GetFraction, but returns the fraction of when the trace left a solid (assuming it started in a solid)
TR_GetDispFlags - Returns the displacement flags of a surface. See DISPSURF_FLAG_* for possible values.
TR_GetSurfaceName - Returns the name of the surface that was hit.
TR_GetSurfaceProps - Returns the surface properties index of the surface that was hit. Not sure if this is very useful alone since it might require other non-TR related functions to be exposed to get any real info from the index.
TR_GetSurfaceFlags - Returns the surface flags of the surface that was hit. See SURF_* for possible values.
TR_GetPhysicsBone - Returns the physics bone index, if one was hit.
TR_AllSolid - Returns whether or not the entire trace was in a solid area.
TR_StartSolid - Returns whether or not the start of the trace was in a solid area.
```

In addition to the new natives, I also changed the existing `TR_EnumerateEntities` native to accept partition flags instead of the `triggers` bool value. This means that the traces made with `TR_EnumerateEntities` are no longer restricted to just "solids or triggers". Instead you can specify any combination of partition flags to use.

The biggest bonus of doing this is that you can now hit non-solid entities with traces. To showcase this (as well as a few of the new natives), here's the output of a little test I made that traces against a non-solid brush, something that was impossible to do previously.
```
Entity: 77 (func_brush)
    Start Pos: 843.38 -185.22 320.03
    Did Hit: true
    All Solid: false
    Start Solid: false
    Fraction Left Solid: 0.000000
    Displacement Flags: 00000000
    Surface Name: DEV/DEV_MEASUREGENERIC01B
    Surface Props: 41
    Surface Flags: 0000000000000000
    Physics Bone: 0
```

For reference, or if anyone else wants to play around with this, here's the plugin I used to do this: https://pastebin.com/xYXvrfa4

Another thing to note is that the partition flag values used in the inc file are a bit different to the original SDK flags. This was done to maintain backwards compatibility so that old plugins that use `true`/`false` still keep their expected behaviour.

### CODE REVIEW

File: extensions/sdktools/trnatives.cpp [hunk 1]
```diff
@@ -114,6 +114,23 @@ enum
 	RayType_Infinite
 };
 
+// For backwards compatibility, old EnumerateEntities functions accepted bool instead of flags
+int TranslatePartitionFlags(int input)
+{
+	if (input == 0)
```
Comment (Headline) [inline extensions/sdktools/trnatives.cpp]: nit: maybe a small switch here instead of `else if`s?

File: plugins/include/sdktools_trace.inc [hunk 2]
```diff
@@ -480,6 +551,71 @@ native void TR_GetEndPosition(float pos[3], Handle hndl=INVALID_HANDLE);
...
+/**
+ * Returns the name of the surface that was hit.
+ *
+ * @param buffer			Buffer to store surface name in
+ * @param maxlen			Maximum length of output buffer
+ * @param hndl			A trace Handle, or INVALID_HANDLE to use a global trace result.
+ * @noreturn
+ * @error				Invalid Handle.
+ */
+native void TR_GetSurfaceName(char[] buffer, int maxlen, Handle hndl=INVALID_HANDLE);
```
Comment (KyleSanderson) [inline plugins/include/sdktools_trace.inc]: The handle or context value must be first to make this into an Object if desired in the future.

Comment (asherkin) [reply plugins/include/sdktools_trace.inc]: Looks like `TR_GetEndPosition` is the only one of the original set that catches us out here.

Comment (SlidyBat) [reply plugins/include/sdktools_trace.inc]: Yeah I specifically moved the handle to last param in a later commit to keep it consistent with `TR_GetEndPosition`. Should I also change `TR_GetStartPosition` or keep that as is?

File: extensions/sdktools/trnatives.cpp [hunk 3]
```diff
@@ -776,6 +806,28 @@ static cell_t smn_TRGetPlaneNormal(IPluginContext *pContext, const cell_t *param
 	return 1;
 }
 
+static cell_t smn_TRGetStartPosition(IPluginContext *pContext, const cell_t *params)
+{
+	sm_trace_t *tr;
+	HandleError err;
+	HandleSecurity sec(pContext->GetIdentity(), myself->GetIdentity());
+
+	if (params[2] == BAD_HANDLE)
```
Comment (KyleSanderson) [inline extensions/sdktools/trnatives.cpp]: Still broken...

File: plugins/include/sdktools_trace.inc [hunk 4]
```diff
@@ -103,7 +103,61 @@
...
+/**
+ * @section Surface flags.
+ */
+ 
+#define	SURF_LIGHT		0x0001		// value will hold the light strength
+#define	SURF_SKY2D		0x0002		// don't draw, indicates we should skylight + draw 2d sky but not draw the 3D skybox
+#define	SURF_SKY		0x0004		// don't draw, but add to skybox
+#define	SURF_WARP		0x0008		// turbulent water warp
+#define	SURF_TRANS		0x0010
+#define SURF_NOPORTAL	0x0020	// the surface can not have a portal placed on it
```
Comment (KyleSanderson) [inline plugins/include/sdktools_trace.inc]: spacing.

File: plugins/include/sdktools_trace.inc [hunk 5]
```diff
@@ -103,7 +103,61 @@
...
+#define	SURF_TRIGGER	0x0040	// FIXME: This is an xbox hack to work around elimination of trigger surfaces, which breaks occluders
+#define	SURF_NODRAW		0x0080	// don't bother referencing the texture
+
+#define	SURF_HINT		0x0100	// make a primary bsp splitter
+
+#define	SURF_SKIP		0x0200	// completely ignore, allowing non-closed brushes
+#define SURF_NOLIGHT	0x0400	// Don't calculate light
```
Comment (KyleSanderson) [inline plugins/include/sdktools_trace.inc]: ditto spacing issue for the entire file.

File: extensions/sdktools/trnatives.cpp [hunk 6]
```diff
@@ -114,6 +114,20 @@ enum
 	RayType_Infinite
 };
 
+// For backwards compatibility, old EnumerateEntities functions accepted bool instead of flags
+int TranslatePartitionFlags(int input)
```
Comment (KyleSanderson) [inline extensions/sdktools/trnatives.cpp]: `static int` if it's just for this file.

**Comment by KyleSanderson** [comment]
Thank you

**Pull request merged**